### PR TITLE
feat(api): Add per-user permissions

### PR DESCRIPTION
**Pull request by dcramer**
This adds a new 'permissions' system which will overtake the standard superuser cases.

For example, there may be a permission which says "you can read data from organizations which you're not a member of", and we'll start using that rather than superuser for many accounts.

### CODE REVIEW

**Comment by dcramer** [comment]
Once this lands ill start instrumenting permissions in the API. Will likely also work towards getting more of _admin into core.

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/auth/access.py
- src/sentry/models/userpermission.py


## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Review by mattrobenolt — APPROVED** [review]

**Comment by dcramer** [comment]
Aside I realize Django has a somewhat equivalent system in core, but I dont want to rely on Django core.

https://docs.djangoproject.com/en/2.0/topics/auth/default/#topic-authorization

**Review by JTCunning — APPROVED** [review]
good stuff.

**Pull request merged**
